src/hooks/useMutex.js

```diff
@@ -0,0 +1,50 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+
+function useMutex() {
+  const queue = React.useRef([]);
+  const lock = React.useRef(false);
+
+  const wait = React.useCallback(() => {
+    const next = queue.current.shift();
+    if (!next) {
+      return;
+    }
+    lock.current = true;
+    return Promise.resolve(next());
+  }, []);
+
+  const signal = React.useCallback(() => {
```

Comment:
`signal` is an operation that lets the mutex know the operation is complete and to release the task. On signal complete, we call `wait` to see if there is another task in the queue that was added while the previous task was executing.